src/units/abilities.cpp

```diff
@@ -1264,14 +1265,66 @@ unit_ability_list attack_type::get_specials_and_abilities(const std::string& spe
 	unit_ability_list spe_list = get_specials(special);
 	if(!spe_list.empty()){
 		spe_list = overwrite_special_checking(special, spe_list, abil_list, "filter_self", true);
-		if(special == "plague" && !spe_list.empty()){
-			return spe_list;
-		}
 		abil_list.append(spe_list);
 	}
 	return abil_list;
 }
 
+unit_ability_list attack_type::get_specials_and_abilities_with_priority(const std::string& special) const
+{
+	//the custom_priority attribute is used for define rules of priority of plague
+	unit_ability_list abil_list = get_specials_and_abilities(special);
+	unit_ability_list temp_list;
+	if(abil_list.empty()){
+		return abil_list;
+	}
+
+	std::vector<double> priorityvec;
+	for(const auto& i : abil_list) {
+		priorityvec.push_back((*i.ability_cfg)["priority"].to_double(0));
+	}
+
+	if(!priorityvec.empty()){
+		double priority = *(std::max_element(priorityvec.begin(), priorityvec.end()));
+		for(const auto& i : abil_list) {
+			if((*i.ability_cfg)["priority"].to_double(0) == priority) {
+				temp_list.emplace_back(i);
+			}
+		}
+	} else {
+		temp_list = abil_list;
+	}
+
+	if(!temp_list.empty()){
+		std::vector<std::string> checking_string;
+		std::vector<config> lists = resources::gamedata->get_special_string_list();
+		if(!lists.empty()){
+			for(const config& str_list : lists) {
+				if(str_list["tag_name"] == special && !str_list["type_list"].empty()){
+					checking_string = utils::split(str_list["type_list"]);
+					break;
+				}
+			}
+		}
+		if(!checking_string.empty()){
+			unit_ability_list priority_list;
+			int i = 0;
+			int n = checking_string.size();
+			while (i>=0 && i<n){
+				for(const auto& j : temp_list) {
+					if((*j.ability_cfg)["type"].str() == checking_string.at(i)){
+						priority_list.emplace_back(j);
+						return priority_list;
```

Comment:
For both `plague` and PR #7003's `damage_type`, the caller only uses the `front()` item of the returned list. If all of the callers will only want one item, it would be clearer to me and less code to return either a `std::optional` or a list that's documented to contain at most one item.

```
If `get_specials_and_abilities(special)` would return an empty list, returns `std::nullopt`.
Otherwise, returns the highest priority item in the list.
If two or more items are tied for highest priority, returns the one that was closer to the start of the list.
```

Reply:
i will document

Reply:
if use of  checking_string approved, can i remove use of priority attribute?

Reply:
i willremain on priority attribute